File: go/sort/leetcode/relative_sort_array.go
package leetcode

import "sort"

// 数组的相对排序
// 使用哈希表，有点复杂的实现。。。
func relativeSortArray1(arr1 []int, arr2 []int) []int {
	m := make(map[int]int)
	for _, v := range arr1 {
		m[v]++
	}

	var res []int
	for _, v := range arr2 {
		for i := 0; i < m[v]; i++ {
			res = append(res, v)
		}
		delete(m, v)
	}

	if len(m) > 0 {
		var temp []int
		for k, v := range m {
			for i := 0; i < v; i++ {
				temp = append(temp, k)
			}
		}
		sort.Ints(temp)
		for _, v := range temp {
			res = append(res, v)
		}
	}
	return res
}

// 使用计数排序的思路
func relativeSortArray2(arr1 []int, arr2 []int) []int {
	temp := make([]int, 1001)
	var res []int
	for _, v := range arr1 {
		temp[v]++
	}

	for _, v := range arr2 {
		for i := 0; i < temp[v]; i++ {
			res = append(res, v)
		}
		temp[v] = 0
	}

	for j := range temp {
		for i := 0; i < temp[j]; i++ {
			res = append(res, j)
		}
	}
	return res
}

File: java/algorithm/sort/leetcode/RelativeSortArray.java
package algorithm.sort.leetcode;

import java.util.Map;
import java.util.TreeMap;

/**
 * @author roseduan
 * 数组的相对排序
 */
public class RelativeSortArray {

    /**
     * 使用计数排序
     */
    public int[] relativeSortArray1(int[] arr1, int[] arr2) {
        int[] temp = new int[1001];
        for (int a : arr1) {
            temp[a]++;
        }

        int i = 0;
        for (int a : arr2) {
            while (temp[a]-- > 0) {
                arr1[i++] = a;
            }
        }

        for (int j = 0; j < temp.length; j++) {
            while (temp[j]-- > 0) {
                arr1[i++] = j;
            }
        }
        return arr1;
    }

    /**
     * 使用一个 TreeMap
     */
    public int[] relativeSortArray2(int[] arr1, int[] arr2) {
        Map<Integer, Integer> map = new TreeMap<>();
        for (int a : arr1) {
            map.put(a, map.getOrDefault(a, 0) + 1);
        }

        int[] res = new int[arr1.length];
        int i = 0;
        for (int a : arr2) {
            for (int j = 0; j < map.get(a); j++) {
                res[i++] = a;
            }
            map.remove(a);
        }

        for (int k : map.keySet()) {
            for (int j = 0; j < map.get(k); j++) {
                res[i++] = k;
            }
        }
        return res;
    }
}
